### PR TITLE
added enpoints for total_enery and energy_list

### PR DESCRIPTION
**Pull request by ianlofs**
/< device_id >/total_energy
Adds support for querying the total energy usage in during a certain time period

/< device_id >/energy_list 
Adds support for querying a list of energy usage where each value is the energy used in some interval between some start and end points.

### CODE REVIEW

File: DB/Vagrantfile [hunk 1]
```diff
@@ -45,7 +46,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   # the path on the guest to mount the folder. And the optional third
   # argument is a set of non-required options.
   # config.vm.synced_folder "../data", "/vagrant_data"
-  config.vm.synced_folder "puppet", "/etc/puppet"
+  config.vm.synced_folder "../../Backend", "/home/vagrant/Backend"
```
Comment (iangudger) [inline DB/Vagrantfile]: Why this change?

File: DB/api/db.py [hunk 2]
```diff
@@ -43,18 +43,27 @@ def query(parsed_url):
...
+	if parsed_url['granularity']:
+		granularity = parsed_url['granularity']
+
+	if parsed_url['total_energy']:
+		results = generate_total_energy(device_id, start_time, end_time, device)
+		return results
+	if parsed_url['energy_list']:
+		results = generate_energy_list(device_id, start_time, end_time, device, granularity)
+		return results
+	else:
```
Comment (iangudger) [inline DB/api/db.py]: This else is not needed because the if returns from the function.

File: DB/api/db.py [hunk 3]
```diff
@@ -43,18 +43,27 @@ def query(parsed_url):
...
-		start_time,
-		end_time,
-		data_type,
-		subset,
-		limit,
-		reverse,
-		device,
-		diff,
-	)
+	if parsed_url['granularity']:
```
Comment (iangudger) [inline DB/api/db.py]: Use the existing style, or if you think this style is better explain why and update the others to match.

File: DB/api/db.py [hunk 4]
```diff
@@ -66,6 +75,80 @@ def query(parsed_url):
...
+
+	:param device_id: The serial number of the device in question
+	:param start_time: The start of the time range for which to query for data
+	:param end_time: The end of the time range for which to query for data
+	:param channel: channel filter
+	"""
+
+	# Initialize parameter list and WHERE clause
+	params = [device_id]
+	query = "SELECT CAST(data as DECIMAL) / 36e5 FROM " + TABLE + " as raw WHERE serial = %s AND type = 'P'"
```
Comment (iangudger) [inline DB/api/db.py]: Why are you dividing by 36e5?

File: DB/api/db.py [hunk 5]
```diff
@@ -66,6 +75,80 @@ def query(parsed_url):
...
+
+	# Generate WHERE clause
+	if start_time and end_time and channel:
+		query += " AND device = %s AND (time = to_timestamp(%s) OR time = to_timestamp(%s));"
+		params.append(channel)
+		params.append(start_time)
+		params.append(end_time)
+		rows = perform_query(query, tuple(params))
+	else:
+		raise Exception("start_time, end_time, or device not provided")
```
Comment (iangudger) [inline DB/api/db.py]: Do this check with the other parameter checks.

File: DB/api/db.py [hunk 6]
```diff
@@ -66,6 +75,80 @@ def query(parsed_url):
...
+	:param channel: channel filter
+	"""
+
+	# Initialize parameter list and WHERE clause
+	params = [device_id]
+	query = "SELECT CAST(data as DECIMAL) / 36e5 FROM " + TABLE + " as raw WHERE serial = %s AND type = 'P'"
+
+	# Generate WHERE clause
+	if start_time and end_time and channel:
+		query += " AND device = %s AND (time = to_timestamp(%s) OR time = to_timestamp(%s));"
```
Comment (iangudger) [inline DB/api/db.py]: We have some gaps in our data. Maybe it would be better to select the range and use MAX and MIN?

File: DB/api/db.py [hunk 7]
```diff
@@ -66,6 +75,80 @@ def query(parsed_url):
...
+	if len(rows) > 1:
+		total_energy = abs(abs(rows[0][0]) - abs(rows[1][0]))
+	else:
+		raise Exception("No power data for the timestamps specified")
+
+	result = '{ total_energy: ' + str(total_energy) + '}'
+	return result
+
+
+def generate_energy_list(device_id, start_time, end_time, channel, granularity):
```
Comment (iangudger) [inline DB/api/db.py]: Why is the function needed? We already have diffing support. Just add chunking to the existing query builder.

File: DB/api/db.py [hunk 8]
```diff
@@ -66,6 +75,80 @@ def query(parsed_url):
...
+	:param device_id: The serial number of the device in question
+	:param start_time: The start of the time range for which to query for data
+	:param end_time: The end of the time range for which to query for data
+	:param channel: channel filter
+	:param granularity: interval for list of values in minutes
+	"""
+
+	# Initialize parameter list and WHERE clause
+	params = [device_id]
+	query = "SELECT time, CAST(lag(data) OVER (ORDER BY time ASC) - data AS DECIMAL)/ 36e5 as delta FROM " + TABLE + " as raw WHERE serial = %s AND type = 'P'"
```
Comment (iangudger) [inline DB/api/db.py]: Again, where did 36e5 come from?

File: DB/api/db.py [hunk 9]
```diff
@@ -66,6 +75,80 @@ def query(parsed_url):
...
+
+	freq = granularity * 60
+	# Generate WHERE clause
+	if channel and granularity:
+		query += " AND device = %s AND time IN (to_timestamp(%s)"
+		params.append(channel)
+		params.append(start_time)
+		start_time += freq
+		while start_time <= end_time:
```
Comment (iangudger) [inline DB/api/db.py]: This is inefficient. Why not do it the way I did in my sample queries with modulo?

Comment (iangudger) [inline DB/api/db.py]: It is important not to let our queries get too long. Long queries will take a long time to build and parse.

File: DB/api/url_parser.py [hunk 10]
```diff
@@ -22,16 +22,31 @@ def parse(url):
...
-	if re.match('^/\d+$', path):
-		query_options['device_id'] = int(path[1:])
-	else:
-		raise Exception("Serial Not Found")
+	device_id_only = re.match('^/(?P<device_id>\d+$)', path)
+	if device_id_only:
+		query_options['device_id'] = int(device_id_only.group('device_id'))
+
+	'''Extract total_energy from URL'''
+	total_energy = re.match('^/(?P<device_id>\d+)/total_energy$', path)
```
Comment (iangudger) [inline DB/api/url_parser.py]: Why not use a parameter for this option? That would be more consistent with our existing style.

File: DB/api/url_parser.py [hunk 11]
```diff
@@ -22,16 +22,31 @@ def parse(url):
...
+	if total_energy:
+		query_options['device_id'] = int(total_energy.group('device_id'))
+		query_options['total_energy'] = True
+
+	'''Extract energy_list from URL'''
+	energy_list = re.match('^/(?P<device_id>\d+)/energy_list$', path)
```
Comment (iangudger) [inline DB/api/url_parser.py]: This one definitely should be a parameter. It is only a little different from the existing diffing functionality.

**Comment by iangudger** [comment]
I have completed my first pass of code review.

File: DB/api/db.py (hunk 4 above)
Comment (ianlofs) [inline DB/api/db.py]: This is the accurate way to get energy from the power data. @aadabi and i verified it last night

File: DB/api/db.py (hunk 3 above)
Comment (ianlofs) [inline DB/api/db.py]: I think its better because all the params are either None or False this way you dont have to do a scan over the list of params every time, not that it matters that much. I'll update.

File: DB/api/url_parser.py (hunk 10 above)
Comment (ianlofs) [inline DB/api/url_parser.py]: I could do it that way, I did it this way because I wanted it to be very explicit, The rest of the api is fairly different what this is

File: DB/api/db.py (hunk 3 above)
Comment (iangudger) [inline DB/api/db.py]: Please update all of them to the new style.

File: DB/api/url_parser.py (hunk 10 above)
Comment (iangudger) [inline DB/api/url_parser.py]: Ok, maybe we can keep this one.

File: DB/api/db.py (hunk 9 above)
Comment (ianlofs) [inline DB/api/db.py]: I couldnt quite figure out what your query was doing, what Im trying to do here is find the difference between data on arbitrary intervals of time, between a start and end point

Comment (iangudger) [inline DB/api/db.py]: My sample queries use a shortcut. Basically, convert the time to a unix timestamp and mod it by the length of the interval.

File: DB/api/db.py (hunk 6 above)
Comment (ianlofs) [inline DB/api/db.py]: I thought about that as well, I guess this is just a first pass at the problem. thought of this
```
select data, to_timestamp(%s) - time difference from data_raw 
where
    to_timestamp(%s) > time 
order by abs(extract(epoch from passed_ts - ts_column))
limit 1;
```

but this way id have to do two queries

Comment (iangudger) [inline DB/api/db.py]: Two queries is fine.

What about this?
```sql
SELECT data
FROM data_raw 
WHERE to_timestamp(%s) > time 
ORDER BY time DESC
LIMIT 1;
```

Comment (iangudger) [inline DB/api/db.py]: Generally if you want two different things you should do two queries. You might want to run them in a single transaction to make sure they are consistent.

Comment (iangudger) [inline DB/api/db.py]: You should avoid having a non-constant number of queries.

File: DB/Vagrantfile (hunk 1 above)
Comment (ianlofs) [inline DB/Vagrantfile]: I found that this was much easier to develop on.

Comment (iangudger) [inline DB/Vagrantfile]: Ok, but I think that breaks automated deployment. Can you keep that change in your local copy?

Comment (iangudger) [inline DB/Vagrantfile]: Maybe keep both?

Comment (ianlofs) [inline DB/Vagrantfile]: will do

**Comment by ianlofs** [comment]
putting together a commit to address your comments @iangudger

File: DB/api/db.py (hunk 9 above)
Comment (ianlofs) [inline DB/api/db.py]: is this the query you were referencing?

```sql
SELECT time, data - lag(data) 
OVER (ORDER BY time) AS delta 
FROM data_raw 
WHERE extract(second from time) = 0 AND device = 'Panel1' 
ORDER BY time 
OFFSET 1;
```

File: DB/api/url_parser.py (hunk 11 above)
Comment (ianlofs) [inline DB/api/url_parser.py]: I agree with you but, my group mates doing the frontend work to graph the data want it in a very specific format which is inconsistent with the rest of the api

File: DB/api/db.py (hunk 9 above)
Comment (iangudger) [inline DB/api/db.py]: Yes. The ```extract(second from time) = 0``` is basically ```extract(epoch from time) % 60 = 0```.

File: DB/api/url_parser.py (hunk 11 above)
Comment (iangudger) [inline DB/api/url_parser.py]: We already have enough complexity. Please try to keep this consistent.

File: DB/api/db.py [hunk 12]
```diff
@@ -184,6 +243,23 @@ def perform_query(query, params):
 	finally:
 		if con:
 			con.close()
+def format_energy_list(rows):
+	"""
+	Formats result set for energy_list query
+
+	:param results:
+	:return:
+	"""
```
Comment (ianlofs) [inline DB/api/db.py]: is this a good compromise for now? I don't think my group members doing the graphing will accept data in any other format at this point. We have a demo on Tuesday that needs to work.

Comment (iangudger) [inline DB/api/db.py]: I like the idea of having energy list be a generic formatting option.

File: DB/api/db.py [hunk 13]
```diff
@@ -129,6 +182,12 @@ def retrieve_within_filters(device_id, start_time, end_time, data_type, subset,
 		if subset:
 			query = write_subsample(query, True)
 
+	if energy_list and device:
+		prefix = "SELECT time, abs(CAST(lag(data) OVER (ORDER BY time DESC) - data AS DECIMAL) / 36e5) FROM " + TABLE + " "
```
Comment (iangudger) [inline DB/api/db.py]: Can we combine this with the diff query logic somehow?

File: DB/puppet/modules/deploy/manifests/init.pp [hunk 14]
```diff
@@ -6,9 +6,9 @@
 CREATE TABLE data_raw (
   serial BIGINT NOT NULL,
   type CHAR(1) NOT NULL,
-  device TEXT NULL,
   data BIGINT NOT NULL,
-  time TIMESTAMP NOT NULL
+  time TIMESTAMP NOT NULL,
+  device TEXT NULL
```
Comment (iangudger) [inline DB/puppet/modules/deploy/manifests/init.pp]: Thanks for making this change. It is better to be consistent about this.

File: DB/api/url_parser.py [hunk 15]
```diff
@@ -47,4 +55,5 @@ def parse(url):
 	if not query_options['device_id']:
 		raise Exception("Not Found")
 
+	print(query_options)
```
Comment (iangudger) [inline DB/api/url_parser.py]: Can we make this more of a logging statement? At the very minimum, add a title. Ex. print("query options:", query_options)

File: DB/api/db.py [hunk 16]
```diff
@@ -184,6 +243,23 @@ def perform_query(query, params):
...
+	:return:
+	"""
+
+	result = "{ data: ["
+
+	for i, row in enumerate(rows):
+		if i > 0 and i < (len(rows) - 1):
+			result += "{ time: " + str(row[0]) + ", energy: " + str(row[1]) + " },"
+		else:
+			result += "{ time: " + str(row[0]) + ", energy: " + str(row[1]) + " }"
```
Comment (iangudger) [inline DB/api/db.py]: We like to do this sort of thing lazily with yield.

File: DB/api/db.py (hunk 12 above)
Comment (iangudger) [inline DB/api/db.py]: Maybe we can make it even more generic by having the "energy:" JSON tag user selectable?

**Comment by iangudger** [comment]
Don't forget to update the documentation.

File: DB/api/db.py (hunk 16 above)
Comment (iangudger) [inline DB/api/db.py]: It avoids having to load the entire result into memory at one time.

File: DB/api/url_parser.py (hunk 15 above)
Comment (ianlofs) [inline DB/api/url_parser.py]: sorry this was a debugging statement ill take it out.

File: DB/api/db.py (hunk 16 above)
Comment (ianlofs) [inline DB/api/db.py]: ok ill update this

**Comment by ianlofs** [comment]
can we leave the generic energy json tag and combining with the diff logic to a separate PR. Id like to get this merged.
